### PR TITLE
Upgrade to Rails 6.0.5

### PR DESCRIPTION
**Pull request by nimmolo**
**Update**: Now passing all tests.

**Original comment**: For me, only two tests fail or error, both in `Api2ControllerTest`. Seems due to `response.body` not being as expected (but why only in these two cases?):
1. `#test_get_observation_with_gps_hidden` - notable that this is only failing on line 364, other assertions seem OK.
```ruby
obs = observations(:unknown_with_lat_long)
get(:observations, params: { id: obs.id, detail: :high, format: :json })
assert_match(/34.1622|118.3521/, @response.body)
```
@response.body shows
```
{"version"=>2.0,
 "run_date"=>"2022-05-26T00:36:03.313Z",
 "query"=>
  "SELECT DISTINCT observations.id FROM `observations` WHERE observations.id = 760938906 ORDER BY observations.id ASC",
 "number_of_records"=>1,
 "number_of_pages"=>1,
 "page_number"=>1,
 "results"=>[""],
 "run_time"=>0.227739763}
```
2. `#test_vote_anonymity` - this appears to be the json is not getting a result, `NoMethodError: private method 'select' called for nil:NilClass`
```ruby
params[:format] = :json
get(:observations, params: params.merge(api_key: rolfs_key.key))
json = JSON.parse(response.body)
votes = json["results"][0]["votes"]
assert_equal(
  "rolf",
  votes.select { |v| v["id"] == rolfs_vote.id }[0]["owner"]["login_name"]
)
```
response.body is similar
```
{"version"=>2.0,
 "run_date"=>"2022-05-26T00:43:41.350Z",
 "user"=>241228755,
 "query"=>
  "SELECT DISTINCT observations.id FROM `observations` WHERE observations.id = 432936842 ORDER BY observations.id ASC",
 "number_of_records"=>1,
 "number_of_pages"=>1,
 "page_number"=>1,
 "results"=>[""],
 "run_time"=>0.769632087}
```

References: 
- [Rails 6.0 Release Notes](https://guides.rubyonrails.org/6_0_release_notes.html)
- [Upgrade Rails from 5.2 to 6.0](https://www.fastruby.io/blog/rails/upgrades/upgrade-rails-from-5-2-to-6-0.html)

### CODE REVIEW

**Comment by nimmolo** [comment]
After analyzing the `@response` object of the two failed tests, I noticed that the `get` in both is actually finding the correct result, but... i the Api2Controller is not putting it in the response body? Below is a sample test response, sorry it's so long. 

What i found burrowing way down through this (here compressed) is the observation in question, right there in the query results. 
```
@request =
  #<ActionController::TestRequest:0x000056054e204438
    @env=
      {"action_controller.instance"=>
        #<Api2Controller:0x00005605482b9000  
          @api=
            #<API2::ObservationAPI:0x000056054e20f040
              @query=
                #<Query::ObservationAll:0x000056054e21b048  
                  @results=
                    {760938906=>
                       #<Observation:0x000056054e2580d8
                         id: 760938906,
                         created_at: Thu, 22 Jul 2010 05:20:00 EDT -04:00,
                         updated_at: Thu, 22 Jul 2010 05:21:00 EDT -04:00,
                         when: Thu, 22 Jul 2010,
                         user_id: 365790011,
                         specimen: false,
                         notes: {:Other=>"unknown_with_lat_long"},
                         thumb_image_id: nil,
                         name_id: 271247339,
                         location_id: 547147019,
                         is_collection_location: true,
                         vote_cache: 0.0,
                         num_views: 0,
                         last_view: nil,
                         rss_log_id: nil,
                         lat: 34.1622,
                         long: -118.3521,
                         where: "Burbank, California, USA",
                         alt: 123,
                         lifeform: " ",
                         text_name: "Fungi",
                         classification: "",
                         gps_hidden: false>},

``` 
contains the desired observation record.
Here is the whole record

```
#<ActionDispatch::TestResponse:0x000056054e2042d0
 @cache_control={},
 @committed=true,
 @cv=
  #<MonitorMixin::ConditionVariable:0x000056054e2041b8
   @cond=#<Thread::ConditionVariable:0x000056054e204190>,
   @monitor=#<ActionDispatch::TestResponse:0x000056054e2042d0 ...>>,
 @header=
  {"X-Frame-Options"=>"SAMEORIGIN",
   "X-XSS-Protection"=>"1; mode=block",
   "X-Content-Type-Options"=>"nosniff",
   "X-Download-Options"=>"noopen",
   "X-Permitted-Cross-Domain-Policies"=>"none",
   "Referrer-Policy"=>"strict-origin-when-cross-origin",
   "Access-Control-Allow-Origin"=>"*",
   "Access-Control-Allow-Headers"=>
    "Origin, X-Requested-With, Content-Type, Accept",
   "Access-Control-Allow-Methods"=>"GET",
   "Content-Type"=>"application/json; charset=utf-8"},
 @mon_count=0,
 @mon_mutex=#<Thread::Mutex:0x000056054e204280>,
 @mon_mutex_owner_object_id=47290392781160,
 @mon_owner=nil,
 @request=
  #<ActionController::TestRequest:0x000056054e204438
   @controller_class=Api2Controller,
   @custom_param_parsers=
    {:xml=>
      #<Proc:0x000056054e204398@/home/vagrant/.rvm/gems/ruby-2.6.6/gems/actionpack-6.0.5/lib/action_controller/test_case.rb:61 (lambda)>},
   @env=
    {"rack.version"=>[1, 3],
     "rack.input"=>#<StringIO:0x000056054e204488>,
     "rack.errors"=>#<StringIO:0x0000560548ba6960>,
     "rack.multithread"=>true,
     "rack.multiprocess"=>true,
     "rack.run_once"=>false,
     "REQUEST_METHOD"=>"GET",
     "SERVER_NAME"=>"example.org",
     "SERVER_PORT"=>"80",
     "QUERY_STRING"=>"",
     "rack.url_scheme"=>"http",
     "HTTPS"=>"off",
     "SCRIPT_NAME"=>"",
     "HTTP_HOST"=>"test.host",
     "REMOTE_ADDR"=>"0.0.0.0",
     "HTTP_USER_AGENT"=>"Rails Testing",
     "action_dispatch.parameter_filter"=>[:password, :password],
     "action_dispatch.redirect_filter"=>[],
     "action_dispatch.secret_key_base"=>
      "475ee75beba3a0c27aa6e1c222645d1524d22c322c974c1c119e4ea89d128317580d1b95ded54d1604b80de91ecf230949a648652674ffd61bc906c44e5ae029",
     "action_dispatch.show_exceptions"=>false,
     "action_dispatch.show_detailed_exceptions"=>true,
     "action_dispatch.logger"=>
      #<Logger:0x000056054abc3248
       @default_formatter=
        #<Logger::Formatter:0x000056054abc31f8 @datetime_format=nil>,
       @formatter=nil,
       @level=3,
       @logdev=
        #<Logger::LogDevice:0x000056054abc31a8
         @dev=#<IO:<STDOUT>>,
         @filename=nil,
         @mon_count=0,
         @mon_mutex=#<Thread::Mutex:0x000056054abc3130>,
         @mon_mutex_owner_object_id=47290364336340,
         @mon_owner=nil,
         @shift_age=nil,
         @shift_period_suffix=nil,
         @shift_size=nil>,
       @progname=nil>,
     "action_dispatch.backtrace_cleaner"=>
      #<Rails::BacktraceCleaner:0x000056054a249b90
       @filters=
        [#<Proc:0x000056054a2496b8@/home/vagrant/.rvm/gems/ruby-2.6.6/gems/activesupport-6.0.5/lib/active_support/backtrace_cleaner.rb:96>,
         #<Proc:0x000056054a249528@/home/vagrant/.rvm/gems/ruby-2.6.6/gems/railties-6.0.5/lib/rails/backtrace_cleaner.rb:16>,
         #<Proc:0x000056054a249500@/home/vagrant/.rvm/gems/ruby-2.6.6/gems/railties-6.0.5/lib/rails/backtrace_cleaner.rb:17>,
         #<Proc:0x000056054a2494d8@/home/vagrant/.rvm/gems/ruby-2.6.6/gems/railties-6.0.5/lib/rails/backtrace_cleaner.rb:18>],
       @root="/vagrant/mushroom-observer/",
       @silencers=
        [#<Proc:0x000056054a249668@/home/vagrant/.rvm/gems/ruby-2.6.6/gems/activesupport-6.0.5/lib/active_support/backtrace_cleaner.rb:100>,
         #<Proc:0x000056054a249618@/home/vagrant/.rvm/gems/ruby-2.6.6/gems/activesupport-6.0.5/lib/active_support/backtrace_cleaner.rb:104>,
         #<Proc:0x000056054a2493c0@/home/vagrant/.rvm/gems/ruby-2.6.6/gems/railties-6.0.5/lib/rails/backtrace_cleaner.rb:19>]>,
     "action_dispatch.key_generator"=>
      #<ActiveSupport::CachingKeyGenerator:0x0000560547dae330
       @cache_keys=
        #<Concurrent::Map:0x0000560547dae308 entries=2 default_proc=nil>,
       @key_generator=
        #<ActiveSupport::KeyGenerator:0x0000560547dae380
         @iterations=1000,
         @secret=
          "475ee75beba3a0c27aa6e1c222645d1524d22c322c974c1c119e4ea89d128317580d1b95ded54d1604b80de91ecf230949a648652674ffd61bc906c44e5ae029">>,
     "action_dispatch.http_auth_salt"=>"http authentication",
     "action_dispatch.signed_cookie_salt"=>"signed cookie",
     "action_dispatch.encrypted_cookie_salt"=>"encrypted cookie",
     "action_dispatch.encrypted_signed_cookie_salt"=>"signed encrypted cookie",
     "action_dispatch.authenticated_encrypted_cookie_salt"=>
      "authenticated encrypted cookie",
     "action_dispatch.use_authenticated_cookie_encryption"=>false,
     "action_dispatch.encrypted_cookie_cipher"=>nil,
     "action_dispatch.signed_cookie_digest"=>nil,
     "action_dispatch.cookies_serializer"=>:marshal,
     "action_dispatch.cookies_digest"=>nil,
     "action_dispatch.cookies_rotations"=>
      #<ActiveSupport::Messages::RotationConfiguration:0x00005605456fe730
       @encrypted=[],
       @signed=[]>,
     "action_dispatch.use_cookies_with_metadata"=>false,
     "action_dispatch.content_security_policy"=>nil,
     "action_dispatch.content_security_policy_report_only"=>false,
     "action_dispatch.content_security_policy_nonce_generator"=>nil,
     "action_dispatch.content_security_policy_nonce_directives"=>nil,
     "rack.session"=>{},
     "rack.session.options"=>
      {:key=>"rack.session",
       :path=>"/",
       :domain=>nil,
       :expire_after=>nil,
       :secure=>false,
       :httponly=>true,
       :defer=>false,
       :renew=>false,
       :sidbits=>128,
       :cookie_only=>true,
       :secure_random=>SecureRandom},
     "HTTP_COOKIE"=>"",
     "PATH_INFO"=>"/api2/observations/760938906.json",
     "action_controller.instance"=>
      #<Api2Controller:0x00005605482b9000
       @_action_has_layout=true,
       @_action_name="observations",
       @_config={},
       @_lookup_context=
        #<ActionView::LookupContext:0x00005605487e4390
         @cache=true,
         @details=
          {:locale=>[:en],
           :formats=>[:json],
           :variants=>[],
           :handlers=>
            [:raw, :erb, :html, :builder, :ruby, :coffee, :jbuilder]},
         @details_key=#<Object:0x000056054d8103f0>,
         @digest_cache=nil,
         @prefixes=["api2", "application"],
         @view_paths=
          #<ActionView::PathSet:0x00005605487e4110
           @paths=
            [#<ActionView::OptimizedFileSystemResolver:0x000056054563d698
              @cache=
               #<ActionView::Resolver::Cache:0x56054563cd60 keys=2 queries=0>,
              @path="/vagrant/mushroom-observer/app/views",
              @pattern=
               ":prefix/:action{.:locale,}{.:formats,}{+:variants,}{.:handlers,}",
              @unbound_templates=
               #<Concurrent::Map:0x000056054563d2d8 entries=2 default_proc=nil>>,
             #<ActionView::OptimizedFileSystemResolver:0x00005605456043e8
              @cache=
               #<ActionView::Resolver::Cache:0x56054563fc40 keys=0 queries=0>,
              @path=
               "/home/vagrant/.rvm/gems/ruby-2.6.6/bundler/gems/graphiql-rails-6b34eb17bf13/app/views",
              @pattern=
               ":prefix/:action{.:locale,}{.:formats,}{+:variants,}{.:handlers,}",
              @unbound_templates=
               #<Concurrent::Map:0x0000560545604398 entries=0 default_proc=nil>>,
             #<ActionView::OptimizedFileSystemResolver:0x00005605456063a0
              @cache=
               #<ActionView::Resolver::Cache:0x5605456060f8 keys=0 queries=0>,
              @path=
               "/home/vagrant/.rvm/gems/ruby-2.6.6/gems/actiontext-6.0.5/app/views",
              @pattern=
               ":prefix/:action{.:locale,}{.:formats,}{+:variants,}{.:handlers,}",
              @unbound_templates=
               #<Concurrent::Map:0x0000560545606378 entries=0 default_proc=nil>>,
             #<ActionView::OptimizedFileSystemResolver:0x00005605455eaab0
              @cache=
               #<ActionView::Resolver::Cache:0x5605455e9b88 keys=0 queries=0>,
              @path=
               "/home/vagrant/.rvm/gems/ruby-2.6.6/gems/actionmailbox-6.0.5/app/views",
              @pattern=
               ":prefix/:action{.:locale,}{.:formats,}{+:variants,}{.:handlers,}",
              @unbound_templates=
               #<Concurrent::Map:0x00005605455ea588 entries=0 default_proc=nil>>]>>,
       @_params=
        <ActionController::Parameters {"detail"=>"high", "id"=>"760938906", "format"=>"json", "controller"=>"api2", "action"=>"observations"} permitted: false>,
       @_request=#<ActionController::TestRequest:0x000056054e204438 ...>,
       @_response=#<ActionDispatch::TestResponse:0x000056054e2042d0 ...>,
       @_response_body=
        ["{\"version\":2.0,\"run_date\":\"2022-05-26T05:31:48.220Z\",\"query\":\"SELECT DISTINCT observations.id FROM `observations` WHERE observations.id = 760938906 ORDER BY observations.id ASC\",\"number_of_records\":1,\"number_of_pages\":1,\"page_number\":1,\"results\":[\"\"],\"run_time\":0.013168806}"],
       @_routes=nil,
       @_url_options=nil,
       @_view_runtime=10.591840000415687,
       @api=
        #<API2::ObservationAPI:0x000056054e20f040
         @action=:observation,
         @detail=:high,
         @errors=[],
         @expected_params=
          {:action=>action: string,
           :detail=>detail: enum (limit=high|low|none),
           :page=>page: integer (default=1),
           :version=>version: float,
           :api_key=>api_key: string,
           :method=>method: string,
           :north=>north: latitude (max latitude),
           :south=>south: latitude (min latitude),
           :east=>east: longitude (max longitude),
           :west=>west: longitude (min longitude),
           :id=>id: integer range list,
           :created_at=>created_at: time range,
           :updated_at=>updated_at: time range,
           :date=>date: date range (when seen),
           :user=>user: user list (observer),
           :name=>name: name list,
           :location=>location: location list,
           :herbarium=>herbarium: herbarium list,
           :herbarium_record=>herbarium_record: herbarium_record list,
           :project=>project: project list,
           :species_list=>species_list: species_list list,
           :confidence=>confidence: confidence (limit=-3..3),
           :is_collection_location=>
            is_collection_location: boolean (is this location where mushroom was found?),
           :gps_hidden=>gps_hidden: boolean (hide exact coordinates?),
           :has_images=>has_images: boolean,
           :has_location=>has_location: boolean,
           :has_name=>has_name: boolean (group or genus or better),
           :has_comments=>has_comments: boolean (limit=true),
           :has_specimen=>has_specimen: boolean,
           :has_notes=>has_notes: boolean,
           :has_notes_field=>
            has_notes_field: string list (is given observation notes template field filled in?),
           :notes_has=>notes_has: string (search within notes),
           :comments_has=>
            comments_has: string (search within comments summary and body),
           :region=>
            region: string (matches locations which end in this, e.g. "California, USA"),
           :include_synonyms=>include_synonyms: boolean,
           :include_subtaxa=>include_subtaxa: boolean,
           :synonyms_of=>synonyms_of: name list,
           :children_of=>children_of: name list},
         @ignore_params={},
         @method=:get,
         @num_results=1,
         @page_number=1,
         @paginator=
          #<MOPaginator:0x000056054e25afb8
           @num_per_page=100,
           @num_total=1,
           @number=1>,
         @params=
          {:detail=>"high",
           :id=>"760938906",
           :action=>:observation,
           :method=>"GET"},
         @query=
          #<Query::ObservationAll:0x000056054e21b048
           @executor=nil,
           @group="",
           @initialized=true,
           @join=[],
           @last_query=
            "\n" +
            "      SELECT DISTINCT observations.id\n" +
            "      FROM `observations` \n" +
            "      WHERE observations.id = 760938906\n" +
            "  ORDER BY observations.id ASC\n" +
            "  LIMIT 0,100\n",
           @num_results=1,
           @order="observations.id ASC",
           @outer_id=nil,
           @params={:where=>["observations.id = 760938906"], :by=>"id"},
           @record=
            #<QueryRecord:0x000056054e22d9f0
             id: nil,
             updated_at: Thu, 26 May 2022 01:31:48 EDT -04:00,
             access_count: 0,
             description:
              "by=$id;flavor=:all;model=:Observation;where=@$observations.id %3D 760938906",
             outer_id: nil>,
           @result_ids=[760938906],
           @results=
            {760938906=>
              #<Observation:0x000056054e2580d8
               id: 760938906,
               created_at: Thu, 22 Jul 2010 05:20:00 EDT -04:00,
               updated_at: Thu, 22 Jul 2010 05:21:00 EDT -04:00,
               when: Thu, 22 Jul 2010,
               user_id: 365790011,
               specimen: false,
               notes: {:Other=>"unknown_with_lat_long"},
               thumb_image_id: nil,
               name_id: 271247339,
               location_id: 547147019,
               is_collection_location: true,
               vote_cache: 0.0,
               num_views: 0,
               last_view: nil,
               rss_log_id: nil,
               lat: 34.1622,
               long: -118.3521,
               where: "Burbank, California, USA",
               alt: 123,
               lifeform: " ",
               text_name: "Fungi",
               classification: "",
               gps_hidden: false>},
           @tables=[],
           @title_args={:type=>:observation, :order=>:sort_by_id},
           @title_tag=:query_title_all_by,
           @where=["observations.id = 760938906"]>,
         @results=
          [#<Observation:0x000056054e2580d8
            id: 760938906,
            created_at: Thu, 22 Jul 2010 05:20:00 EDT -04:00,
            updated_at: Thu, 22 Jul 2010 05:21:00 EDT -04:00,
            when: Thu, 22 Jul 2010,
            user_id: 365790011,
            specimen: false,
            notes: {:Other=>"unknown_with_lat_long"},
            thumb_image_id: nil,
            name_id: 271247339,
            location_id: 547147019,
            is_collection_location: true,
            vote_cache: 0.0,
            num_views: 0,
            last_view: nil,
            rss_log_id: nil,
            lat: 34.1622,
            long: -118.3521,
            where: "Burbank, California, USA",
            alt: 123,
            lifeform: " ",
            text_name: "Fungi",
            classification: "",
            gps_hidden: false>],
         @user=nil,
         @version=2.0>,
       @browser=
        #<Browser::Unknown:0x000056054d375520
         @accept_language_raw="",
         @bot=
          #<Browser::Bot:0x000056054d3753e0
           @browser=
            #<Browser::Unknown:0x000056054d357cc8
             @accept_language_raw="",
             @ua="rails testing">,
           @ua="rails testing">,
         @ua="Rails Testing">,
       @marked_for_same_origin_verification=false,
       @rendered_format=
        #<Mime::Type:0x0000560545417850
         @hash=4514549933178281575,
         @string="application/json",
         @symbol=:json,
         @synonyms=["text/x-json", "application/jsonrequest"]>,
       @start_time=Thu, 26 May 2022 01:31:48 EDT -04:00,
       @user=nil>,
     "action_dispatch.request.path_parameters"=>
      {:id=>"760938906",
       :format=>"json",
       :controller=>"api2",
       :action=>"observations"},
     "action_dispatch.request.flash_hash"=>
      #<ActionDispatch::Flash::FlashHash:0x000056054e20afe0
       @discard=#<Set: {}>,
       @flashes={},
       @now=nil>,
     "action_dispatch.request.content_type"=>nil,
     "action_dispatch.request.request_parameters"=>{},
     "rack.request.query_string"=>"detail=high",
     "rack.request.query_hash"=>{"detail"=>"high"},
     "action_dispatch.request.query_parameters"=>{"detail"=>"high"},
     "action_dispatch.request.parameters"=>
      {"detail"=>"high",
       "id"=>"760938906",
       "format"=>"json",
       "controller"=>"api2",
       "action"=>"observations"},
     "action_dispatch.request.formats"=>
      [#<Mime::Type:0x0000560545417850
        @hash=4514549933178281575,
        @string="application/json",
        @symbol=:json,
        @synonyms=["text/x-json", "application/jsonrequest"]>]},
   @filtered_env=nil,
   @filtered_parameters=
    {"detail"=>"high",
     "id"=>"760938906",
     "format"=>"json",
     "controller"=>"api2",
     "action"=>"observations"},
   @filtered_path=nil,
   @fullpath="/api2/observations/760938906.json?detail=high",
   @headers=
    #<ActionDispatch::Http::Headers:0x000056054e209e60
     @req=#<ActionController::TestRequest:0x000056054e204438 ...>>,
   @ip="0.0.0.0",
   @method="GET",
   @original_fullpath=nil,
   @port=80,
   @protocol="http://",
   @remote_ip="0.0.0.0",
   @request_method="GET",
   @variant=[]>,
 @sending=false,
 @sent=true,
 @status=200,
 @stream=
  #<ActionDispatch::Response::Buffer:0x000056054e2b19a8
   @buf=
    ["{\"version\":2.0,\"run_date\":\"2022-05-26T05:31:48.220Z\",\"query\":\"SELECT DISTINCT observations.id FROM `observations` WHERE observations.id = 760938906 ORDER BY observations.id ASC\",\"number_of_records\":1,\"number_of_pages\":1,\"page_number\":1,\"results\":[\"\"],\"run_time\":0.013168806}"],
   @closed=false,
   @response=#<ActionDispatch::TestResponse:0x000056054e2042d0 ...>,
   @str_body=
    "{\"version\":2.0,\"run_date\":\"2022-05-26T05:31:48.220Z\",\"query\":\"SELECT DISTINCT observations.id FROM `observations` WHERE observations.id = 760938906 ORDER BY observations.id ASC\",\"number_of_records\":1,\"number_of_pages\":1,\"page_number\":1,\"results\":[\"\"],\"run_time\":0.013168806}">>

**Comment by nimmolo** [comment]
Most of the brakeman issues are "Unprotected mass assignment" & "Mass assignment is not restricted using attr_accessible"
I'd like to approach fixing these in PRs to master, prior to this upgrade.

**Comment by pellaea** [comment]
I'll take a look at this.

**Comment by pellaea** [comment]
I'm looking at the second API2 failure, and I've narrowed it down to json.results(array, partial: blah) not working the way it should.  This is weird, because presumably it must work correctly elsewhere in the test suite!  I'm trying to figure out why this particular one is failing.  It definitely renders app/views/api2/_observation.json.builder, and it's definitely not crashing or anything inside... it just seems to be throwing away the result.  I'm kinda stumped!

**Comment by pellaea** [comment]
Strike that, sorry, collection doesn't work.

**Comment by pellaea** [comment]
Yikes.  "json" has been changed, possibly radically, in this version.  I cannot even figure out what class it is.  It claims to be "nil", but of course, nil.class.name == NilClass not nil, so it must be overriding "class".  How can one determine the class of an instance that overrides class?!!

It seems clear to me that the way rendering in JSON is *supposed* to work in rails is via some magical to_json method.  I think.  But I don't see how to have it pass the detail level argument if I'm not the one actually calling the to_json method.  Dammit, I would really prefer not to have to rewrite the whole goddamned thing.  This is pissing me off.

**Comment by pellaea** [comment]
It appears to be JBuilder -- see: https://github.com/rails/jbuilder

**Comment by pellaea** [comment]
Alright, I'm ready to pronounce this is a bug in Jbuilder, not our code.  I cannot get collections to render anything at all, no matter how much I simplify the partial.  Do we believe it is our responsibility to track down and fix errors in public external gems?

**Comment by nimmolo** [comment]
Thank you for checking this out! That’s all i need, it’s a huge help already.
i’ll start researching the Jbuilder situation.

nimmo

**Comment by nimmolo** [comment]
Okay, I now think these 75 Brakeman warnings are false positives. 

We already handle mass assignment the recommended way — not using `attr_accessible`, but in our controllers with methods like this, in AccountController, using `params.require` and `params.permit`.
```ruby
  private

  def initialize_new_user
    now = Time.zone.now
    @new_user.attributes = {
      created_at: now,
      updated_at: now,
      last_login: now,
      admin: false,
      layout_count: 15,
      mailing_address: "",
      notes: ""
    }.merge(params.require(:new_user).permit(:login, :name, :theme,
                                             :email, :email_confirmation,
                                             :password, :password_confirmation))
  end
```

**Comment by nimmolo** [comment]
> Oh wait, no, we don't actually ever test it on a nonempty result anywhere else in that test suite! :) Okay, there's our problem. json.results(array, partial: blah) has apparently changed. We just need to figure out how...

Jason - Should I add a "test on a nonempty result" to Api2ControllerTest? 
Would I just get a record and assert that the controller returns an expected result?

File: Gemfile.lock [hunk 1]
```diff
@@ -18,64 +18,78 @@ GEM
...
+    actionmailer (6.0.5)
+      actionpack (= 6.0.5)
+      actionview (= 6.0.5)
+      activejob (= 6.0.5)
       mail (~> 2.5, >= 2.5.4)
       rails-dom-testing (~> 2.0)
-    actionpack (5.2.7.1)
-      actionview (= 5.2.7.1)
-      activesupport (= 5.2.7.1)
+    actionpack (6.0.5)
```
[the remark this answers is not in the document]
Comment (nimmolo) [reply Gemfile.lock:38]: Supposedly [fixed](https://discuss.rubyonrails.org/t/cve-2021-22881-possible-open-redirect-in-host-authorization-middleware/77130) in 6.0.3.5

**Comment by pellaea** [comment]
Actually that test *does* test a nonempty result.  It is the only one that
does, right now, but it's enough I suppose.

**Comment by nimmolo** [comment]
I've never done a Rails version upgrade, so i don't know all the things to check for. 
But this seems to have been pretty minor so far, thanks to Jason diagnosing what was amiss with his eponymous method "json". We don't need to switch to Webpack, and we don't use websockets anywhere as far as I know. Can it be that we're ready?

There are a flurry of config changes listed [here](https://railsdiff.org/5.2.7.1/6.0.5), but i'm not sure any of them apply to us. Have a look, it's a handy diff of files that would be changed by the task `rails app:update`, which I **did not run** for the above reason. They can be inserted line by line if we need them.

The ones that caught my notice:
- `config/environments/development.rb`, line 5 (enable fragment cache logging)
- `config/environments/test.rb`, lines 10, 11 and 19 (other caching behavior)
- `test/test_helper.rb`(Run tests in parallel with specified workers)

**Comment by pellaea** [comment]
You have a better eye than I.  I could only find the thing about
parallelizing tests.  You seem to have lots of energy and time on your
hands (haha!) -- maybe it would be fun to try running the tests with that
turned on for kicks?

Anyway, if the tests all pass, and there were very few (any?) major issues
that needed to be addressed with the upgrade, then I'm fine with merging it
and pushing it out.  I'll do the push if you do the merge.
-Jason

**Comment by nimmolo** [comment]
OK! Removed extranea, preparing for merge.

**Comment by pellaea** [comment]
Awesome.  Thanks as always.

**Pull request merged**